Add orderBy & first parameters to groupedAggregates

faridnsh opened the pull request:
More context on #75. Closes #75 

Basically adds `orderBy`, `first` and `last` parameters to `groupedAggregates`. `orderBy` accepts any possible field aggregations. Note that I still don't have a very good understanding of the project, this PR is party AI generated and partly from copy pasting various concepts around, would appreciate deep review from someone who understands the code better to see if there can better ways.

## Performance impact

Slight performance impact in startup time due to added(18 entries per numerical fields in AggregatesOrderBy enum)

## Security impact

Shouldn't be.

## Checklist

- [ ] My code matches the project's code style and `yarn lint:fix` passes. - I'm working on estlint to run properly
- [x] I've added tests for the new feature, and `yarn test` passes.
- [x] I have detailed the new feature in the relevant documentation.
- [x] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists). - File doesn't exist
- [x] If this is a breaking change I've explained why. - not a breaking change

faridnsh committed:
Add orderBy, first and last parameters to groupedAggregates

Closes #75

benjie committed:
Merge branch 'v5' into orderBy

benjie commented on README.md:
```diff
@@ -271,15 +271,30 @@ See
+  trailing `n` groups.  
+
+Always pair `first` or `last` with an explicit `orderBy` so PostgreSQL can
+deterministically rank the groups before trimming them.
```
Suggest that we remove `last` - if they need that they should do the inverse order.

benjie commented on README.md:
````diff
@@ -317,6 +334,9 @@ query AverageGoalsOnDaysWithAveragePointsOver200 {
 ```
 
+When using `last`, be sure to supply an `orderBy` so the database can produce a
+deterministic ordering before the tail slice is applied.
````
We should default to ordering by the group by clauses; I didn't realise we didn't do that already.

benjie commented on __tests__/queries/groupedAggregatesOrderBy.test.ts:
```diff
@@ -0,0 +1,31 @@
+        bottom: groupedAggregates(
+          groupBy: [PLAYER_ID]
+          orderBy: [SUM_POINTS_ASC]
+          last: 2
```
```suggestion
          orderBy: [SUM_POINTS_ASC]
          first: 2
```

benjie commented on README.md:
```diff
@@ -317,6 +334,9 @@ query AverageGoalsOnDaysWithAveragePointsOver200 {
 
+When using `last`, be sure to supply an `orderBy` so the database can produce a
+deterministic ordering before the tail slice is applied.
+
```
```suggestion
```

benjie commented on __tests__/queries/__snapshots__/groupedAggregatesOrderBy.test.ts.snap:
```diff
@@ -0,0 +1,56 @@
+}
+`;
+
+exports[`GroupedAggregatesOrderBy: sql 1`] = `
+[
+  "select
+  (coalesce(sum(__match_stats__."points"), '0'))::text as "0",
+  __match_stats__."player_id"::text as "1"
+from "test"."match_stats" as __match_stats__
+group by __match_stats__."player_id"
+order by coalesce(sum(__match_stats__."points"), '0') desc
+limit 2;",
+]
+`;
```
:thinking: Where's the other query...?

faridnsh replied:
My understanding is that when switching to last with reverse ordering, it became the same exact query for the database and it was deduplicated so it was never executed.

benjie replied:
I'm often caught out by Grafast being cleverer than me :laughing:

benjie commented on src/AddConnectionGroupedAggregatesPlugin.ts:
```diff
@@ -151,6 +154,75 @@ const Plugin: GraphileConfig.Plugin = {
+                      },
+                    []
+                  ),
+                },
```
```suggestion
```

benjie reviewed (changes requested):
This is really thorough, I like it! I wonder how we avoid type explosion with this - perhaps we should have an opt-in to the feature? I don't think index behaviours can save us here. This plugin already adds a lot of types to the schema :sweat_smile: 

For now let's drop the `last` argument.

faridnsh commented:
> perhaps we should have an opt-in to the feature?

I think definitely a good idea and safe default. I'm making it happen through behaviours.

faridnsh committed:
Add per-aggregate behaviors for groupedAggregates orderBy

Enhances the groupedAggregates:orderBy behavior system to support
per-aggregate scoping, providing more granular control over which
aggregates can be used for ordering grouped results.

Changes:
- Add per-aggregate behavior declarations (sum:, min:, max:, average:,
  distinctCount:, stddevSample:, stddevPopulation:, varianceSample:,
  variancePopulation:) for groupedAggregates:orderBy
- Simplify entity behavior registration by removing redundant array
  wrapper
- Remove manual hierarchical behavior checking logic in favor of
  PostGraphile's built-in behavior precedence system
- Inline resource validation check for better encapsulation

Documentation:
- Document groupedAggregates:orderBy behavior with explanation of why
  it's disabled by default (18 enum values per attribute = schema bloat)
- Add examples showing three levels of granularity: table-level,
  column-level, and per-aggregate scoping
- Follow existing documentation style for consistency

This allows users to selectively enable only the specific aggregates
they need (e.g., +sum:attribute:aggregate:groupedAggregates:orderBy)
rather than all 9 aggregates, reducing schema bloat while maintaining
flexibility.

faridnsh committed:
Remove 'last' argument from groupedAggregates field

The 'last' argument is the opposite of 'first' and adds unnecessary
complexity for grouped aggregates. Users can achieve the same result
by reversing the orderBy direction and using 'first'.

Changes:
- Remove 'last' argument from groupedAggregates field definition
- Update README to only document 'first' argument
- Update test query to remove usage of 'last' argument
- Regenerate snapshots with updated schema

This simplifies the API while maintaining all necessary functionality.

faridnsh committed:
Add default ordering by GROUP BY columns for groupedAggregates

When no explicit orderBy is specified, groupedAggregates now returns
results ordered by the GROUP BY columns in ascending order. This ensures
deterministic, predictable results instead of database-dependent ordering.

Changes:
- Modify groupBy enum value apply functions to add default ORDER BY
- Apply to both direct attributes and derivative group-by specs
- Update README to document default ordering behavior
- Regenerate test snapshots with new ORDER BY clauses in SQL

This provides a better developer experience with consistent, sorted
results by default while still allowing custom ordering via orderBy
argument when needed.

faridnsh committed:
Add tests for per-aggregate groupedAggregates orderBy behaviors

Adds comprehensive tests validating that per-aggregate behavior scoping
works correctly for groupedAggregates orderBy functionality.

Tests verify:
- Enabling only SUM aggregate for orderBy (excludes MIN, MAX, AVERAGE, etc.)
- Enabling multiple specific aggregates (SUM + AVERAGE)
- Proper behavior precedence (more specific beats less specific)

Example usage shown in tests:
  -attribute:aggregate:groupedAggregates:orderBy +sum:attribute:aggregate:groupedAggregates:orderBy

This allows users to selectively enable only the aggregates they need,
reducing schema bloat from 18 enum values per attribute down to 2 per
enabled aggregate type.

faridnsh committed:
Fix README example for per-aggregate behavior usage

Corrects the per-aggregate behavior example to show that the generic
behavior must be disabled first before enabling specific aggregates.

Before (incorrect):
  +sum:... +average:...

After (correct):
  -attribute:aggregate:groupedAggregates:orderBy +sum:... +average:...

Without the initial `-attribute:...`, all aggregates would be enabled
due to behavior precedence, defeating the purpose of per-aggregate
scoping. Adds explanatory note to prevent confusion.

faridnsh committed:
Fix yarn lint:fix

faridnsh commented:
Implemented all suggestions, with also some tests for the per-aggregate behaviours, let me know if they make sense and it's not too much testing.

benjie commented:
This looks great from a quick glance! I'm going to add it to my post-v5.0.0 review list, so with the winter break coming up and me at a conference next week it might be a couple months until I get around to reviewing and merging it, hopefully you can make do with your own fork until then :crossed_fingers:
